### ISSUE TITLE
flirt: consider using rizin/sigdb signatures

### ISSUE THREAD

**Issue by williballenthin** [mandiant/capa#1094]
https://github.com/rizinorg/sigdb

**Comment by r0ny123** [comment]
We can also use this for floss.

**Comment by Aayush-Goel-04** [comment]
@williballenthin ,
I have looked into adding `.sig` files. The paths to sig files are being loaded by `viv_utils.flirt.load_flirt_signature` we can make `sigs` folder a submodule linked to https://github.com/rizinorg/sigdb . We can add path to sigs folder as default while running capa if not specified by user.

**Comment by williballenthin** [comment]
that sounds good.

prior to doing this, would you investigate the coverage provided by the rizin sig files versus what we currently distribute? you can match FLIRT signatures against a collection of PE files, perhaps around 100, and see how often each symbol matches. then, we should be able to decide if we want to commit to rizin.

**Comment by Aayush-Goel-04** [comment]
> see how often each symbol matches.

Do you mean the identifying how often library functions matches? I think the script to check the coverage of functions can be generated by making some modifications in `scripts/match-function-id.py`?
thoughts @williballenthin

**Comment by williballenthin** [comment]
i can also provide a collection of random files or you can use the files in capa-testfiles. the second idea might be a bit easier since they're already available.

**Comment by williballenthin** [comment]
what im most interested in is if the rizin database matches approximately the same number (or more) functions than the siglib databases. its reassuring to see about the same names in the results above! i don't think its important to investigate every difference - just the approximate total counts and coverage.

i think "?" is possibly used to indicate there were matches but its not clear which one (ambiguous match).

i dont quite understand your question, though. did the above explanation help? or if you need a different response, can you rephrase the question?

**Comment by williballenthin** [comment]
yeah please include all the testfiles, if possible

**Comment by Aayush-Goel-04** [comment]
@williballenthin 
running 
`python3 scripts/match-function-id.py  tests/data/6cc148363200798a12091b97a17181a1.exe_ --signature sigs/1_flare_msvc_rtf_32_64.sig` 
gives error at function `0x1401d4e60: 'NoneType' object is not subscriptable`. Can you think of whats causing error here?
```
Traceback (most recent call last):
  File "/Users/ayush.goel/Documents/GitHub/capa/scripts/match-function-id.py", line 134, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/ayush.goel/Documents/GitHub/capa/scripts/match-function-id.py", line 126, in main
    name = viv_utils.flirt.match_function_flirt_signatures(analyzer.matcher, vw, function)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/viv_utils/flirt.py", line 188, in match_function_flirt_signatures
    loc_va = vw.getLocation(ref_va)[vivisect.const.L_VA]
             ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

**Comment by williballenthin** [comment]
looks like getLocation is returning None when we expect it to be a tuple. this needs a fix in viv-utils. you're welcomed to propose this or i can do it tomorrow. in the meantime, maybe update the script to catch such an exception?

**Comment by Aayush-Goel-04** [comment]
<img width="785" alt="image" src="https://github.com/mandiant/capa/assets/81844215/bf26a4a4-cc81-46f9-b4e7-10e1ec5209d1">
Processed a total of 219 .exe_ files from capa-testfiles. For 102 cases rizin/sigdb had more matches that capa sigs/ ones.

**Comment by Aayush-Goel-04** [comment]
Here the csv file contanining number of symbol matches for each with both capa and rizin sigs.
[results_for_rizin&capa_sigs.csv](https://github.com/mandiant/capa/files/11739399/results_for_rizin.capa_sigs.csv)

**Comment by Aayush-Goel-04** [comment]
> looks like getLocation is returning None when we expect it to be a tuple. this needs a fix in viv-utils.

Done in PR https://github.com/williballenthin/viv-utils/pull/116

**Comment by Aayush-Goel-04** [comment]
<img width="993" alt="image" src="https://github.com/mandiant/capa/assets/81844215/3ed4b454-380b-44f5-8a13-07c982b83c1f">
For all 219 exe files in capa-testfiles processed rizin gave more matches for 101 files. In rest cases capa gave better results. Below imgae shows difference in matches for all files.
<img width="992" alt="image" src="https://github.com/mandiant/capa/assets/81844215/8c705bee-a985-4156-9b7c-918b2a90f07e">
[allMatches.xls](https://github.com/mandiant/capa/files/11743122/allMatches.xls)

**Comment by Aayush-Goel-04** [comment]
@williballenthin 
Based on above results it would be better to stick to current Capa sigs.
What are your thoughts !

**Comment by mr-tz** [comment]
Interesting results. Do you have insight into which files rizin handles better than the capa signatures? Maybe we can leverage a subset of the rizin rules?

**Comment by Aayush-Goel-04** [comment]
In the excel file attached above it mentions which files rizin handles better. Currently, I don't know a way to classify the PE files used for testing. Do you have any suggestions of how I can classify files ?

**Comment by mr-tz** [comment]
detect it easy could shed some light on compiler/linker versions

**Comment by Aayush-Goel-04** [comment]
Sharing compiler and linkers results found using Detect it easy
[updatedMatches.xlsx](https://github.com/mandiant/capa/files/11860051/updatedMatches.xlsx)

I wasn't able to detect any patterns in files which rizin handles better. @mr-tz @williballenthin Could you share your views if u found any patterns based on results in above excel file.

**Comment by mr-tz** [comment]
Cool, thanks for the research here. Seems like rizin does slightly better on

```
Microsoft Linker(14.11, Visual Studio 2017 15.3*)[Console32,console]
Microsoft Linker(14.0, Visual Studio 2015 14.0*)[Console64,console]
Microsoft Linker(14.0, Visual Studio 2015 14.0*)[GUI32]
Microsoft Linker(14.26, Visual Studio 2019 16.6*)[Console32,console]
Microsoft Linker(14.0, Visual Studio 2015 14.0*)[GUI64]
Microsoft Linker(14.0, Visual Studio 2015 14.0*)[Console32,console]
Microsoft Linker(14.32, Visual Studio 2022 17.2*)[GUI64]
```

But overall, our signatures seem to do pretty well.

**Comment by Aayush-Goel-04** [comment]
@mr-tz, thanks for reviewing the results.
Let me know if you have any further thoughts or if there's anything else regarding this.

**Comment by mr-tz** [comment]
@williballenthin, what do you think of these results, can we close this issue (for now)?

**Comment by williballenthin** [comment]
yeah, i agree, lets keep using the signatures that we have; however, if they become noticably out of data and/or rizin introduces other relevant signatures, lets consider if it become worthwhile to switch over.

@Aayush-Goel-04 thank you very much for taking the time to do this data exploration. although it didn't lead to a merged PR, this is a better outcome - no additional work. i appreciated the way you collected data and presented the results, sharing raw information when we asked for it. 🙇🏼‍♂️
